File: 12/12.clj
(defn divisors [n]
  (let [limit (int (Math/sqrt n))
        divisible? (fn [n d] (zero? (mod n d)))]
    (->> (range 1 (inc limit))
         (filter #(divisible? n %))
         (mapcat #(vector % (/ n %)))
         set)))

(defn triangle-numbers
  ([] (triangle-numbers 1 2))
  ([n i] (lazy-seq (cons n (triangle-numbers (+ n i) (inc i))))))

(->> (triangle-numbers)
     (filter (fn [n] (> (count (divisors n)) 500)))
     first
     (print))

File: 12/12.go
package main

import (
	"fmt"
	"math"
)

func divisors(n int) []int {
	var divisors []int
	limit := int(math.Sqrt(float64(n)))
	for d := 1; d <= limit; d++ {
		if n % d == 0 {
			divisors = append(divisors, d)
			if d != n / d {
				divisors = append(divisors, n / d)
			}
		}
	}

	// fmt.Printf("%v: %v\n", n, divisors)
	return divisors
}

func main() {
	n := 1
	i := 1
	for len(divisors(n)) < 500 {
		i += 1
		n += i
	}
	fmt.Println(n)
}
